source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+// Implements the interface required by
+// https://quiche.googlesource.com/quiche/+/refs/heads/master/quic/platform/api/quic_mem_slice.h
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a buffer fragment and the buffer of given length the fragment points to with the
+  // guarantee that buffer and fragment are both aligned according to max_align_t.
+  static Envoy::Buffer::BufferFragmentImpl*
+  allocateBufferAndFragment(QuicBufferAllocator* allocator, size_t length);
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length);
+
+  // Constructs a QuicMemSliceImpl from a Buffer::Instance with first |length| bytes in it.
+  // Data will be moved from |buffer| to this mem slice.
+  // Prerequisite: |buffer| has at least |length| bytes of data and not empty.
+  explicit QuicMemSliceImpl(Envoy::Buffer::Instance& buffer, size_t length);
+
+  QuicMemSliceImpl(const QuicMemSliceImpl& other) = delete;
+  // Move constructors. |other| will not hold a reference to the data buffer
+  // after this call completes.
+  QuicMemSliceImpl(QuicMemSliceImpl&& other) noexcept { *this = std::move(other); }
+
+  QuicMemSliceImpl& operator=(const QuicMemSliceImpl& other) = delete;
+  QuicMemSliceImpl& operator=(QuicMemSliceImpl&& other) noexcept {
+    if (this != &other) {
+      single_slice_buffer_.move(other.single_slice_buffer_);
+    }
+    return *this;
+  }
+
+  // Below methods implements interface needed by QuicMemSlice.
+  void Reset() { single_slice_buffer_.drain(length()); }
+
+  // Returns a char pointer to the one and only slice in buffer.
+  const char* data() const;
+
+  size_t length() const { return single_slice_buffer_.length(); }
+  bool empty() const { return length() == 0; }
+
+private:
+  // Used to align both fragment and buffer at max aligned address.
+  struct BufferFragmentBundle {
+    // Wrap fragment in a nested struct so that it can be padded according to
+    // its alignment requirement: max_align_t. This ensures buffer to start at a
+    // max aligned address.
+    struct {
```

Comment:
I still don't see why you need this. The C++ compiler knows about the alignment requirements of Envoy::Buffer::BufferFragmentImpl. I think this will all be correct if you just allocate the BufferFragmentImpl directly.

If you are saying that the alignment imposed by the C++ compiler is wrong I'd like to understand why. If you are saying that the default alignment is correct but suboptimal, please explain in more detail. I see the comment above but it doesn't explain why you need to override the compiler's memory model.

Reply:
We have a very similar conversation to this in https://github.com/envoyproxy/envoy/pull/6732#discussion_r279602553 (CC @lizan @jplevyak ). I think this is a sign that things are more complicated than they need to be here. Can we make this safer by construction?

Reply:
Chime in since this is my suggestion to Dan.

The reason for the alignment is very simple: If we new() the BufferFragmentImpl and buffer separately, we'll get two pointers aligned to std::max_align_t.  The nested struct+alignas in this structure make sure the two addresses are still aligned to std::max_align_t, even if they are allocated together. As you said, this is not necessary. (unless someone casts a pointer to buffer into a pointer to some object on a processor that doesn't support unaligned access, which is very hacky and should never happen.)

The reason for combining them is to reduce the number of allocations, see my previous comments at [here](https://github.com/envoyproxy/envoy/pull/6400#discussion_r270143912), [here](https://github.com/envoyproxy/envoy/pull/6400#discussion_r271452624) and [here](https://github.com/envoyproxy/envoy/pull/6400#discussion_r271874854).

If performance is not a concern here, I'm ok to remove the nested struct+alignas, or if you prefer, even go back to the original version.

Reply:
That was not my suggestion. My suggestion was simply to define a nested structure and let the compiler take care of alignment, but maybe I'm missing something.

Reply:
Sorry -- rewriting my comment.

The pattern of having a structure with a variable-size buffer nested in another structure is used multiple times in Envoy, without having to mess with alignment in such a finicky way.

Also, abstraction boundaries are at issue here.

I agreed with your comment about using 1 instead of 2 allocations, but I hope we can do this with more straightforward code.

Reply:
QuicMemSlice::data() is used in hot path of quic write path. Accessing the memory needs to be very efficient. The code definitely works without all the fragment_with_padding_ magic. But I think we will eventually need similar optimization in production code.

Reply:
Can we grab a whiteboard when I'm in back in the office Monday? I believe fully in the aggregated memory block and I don't understand why you need this magic to achieve that.